[quant][onnx] Add support to convert max_pool2d quantized pytorch op to C2

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34629 [quant][onnx] Support conversion of quantized sigmoid operator from pytorch to caffe2
* **#33945 [quant][onnx] Add support to convert max_pool2d quantized pytorch op to C2**

Summary:
Add mapping for this operator in symbolics

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_max_pool2d

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20433681](https://our.internmc.facebook.com/intern/diff/D20433681)

supriyar committed:
[quant][onnx] Add support to convert max_pool2d quantized pytorch op to C2

Summary:
Add mapping for this operator in symbolics

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_max_pool2d

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][onnx] Add support to convert max_pool2d quantized pytorch op to C2"


Summary:
Add mapping for this operator in symbolics

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_max_pool2d

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -154,6 +154,24 @@ def upsample_nearest2d(g, input, output_size, align_corners=None, scales_h=None,
+        "Y_scale_f": input.node()["Y_scale"],
+        "Y_zero_point_i": input.node()["Y_zero_point"],
+    }
+    input = nchw2nhwc(g, input)
```
Is there a way to check if it is already in NHWC? That way we could add a fast pass here

supriyar replied:
The way this conversion flow is written, it always adds these permutes before and after conv ops. Else branches in the graph become hard to handle. Once we convert the model we can use a script to remove the unnecessary permutes from the graph.

z-a-f reviewed (approved):
LGTM

raghuramank100 commented on test/onnx/test_pytorch_onnx_caffe2_quantized.py:
```diff
@@ -204,5 +204,19 @@ def forward(self, x, y):
+                return self.dequant(res)
+
+        x = np.random.rand(1, 2, 8, 8).astype("float32")
+        self.generic_test(QMaxPool2dModule(), (x,), input_names=["x"])
```
Can we check for an exact match here for maxpool?  Also, should we test for different values of stride/padding?

houseroad replied:
If we want to check for exact match for maxpool, then I think we need to add parameter decimal to generic_test

houseroad reviewed (commented):
Overall, looks good.

supriyar committed:
Update on "[quant][onnx] Add support to convert max_pool2d quantized pytorch op to C2"


Summary:
Add mapping for this operator in symbolics

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_max_pool2d

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20433681](https://our.internmc.facebook.com/intern/diff/D20433681)

[ghstack-poisoned]

raghuramank100 commented on test/onnx/test_pytorch_onnx_caffe2_quantized.py:
```diff
@@ -204,5 +204,19 @@ def forward(self, x, y):
+                return self.dequant(res)
+
+        x = np.random.rand(1, 2, 8, 8).astype("float32")
+        self.generic_test(QMaxPool2dModule(), (x,), input_names=["x"], decimal=0)
```
We should be using a larger value for decimal, from the docs:
https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.testing.assert_almost_equal.html
abs(desired-actual) < 1.5 * 10**(-decimal)

Can we use decimal of 5?
